Add/https

tylerburleigh opened the pull request:
**Description of the Pull Request (PR):**

This simplifies the process of SSL certificate generation.

**This fixes or addresses the following GitHub issues:**

- Ref: #


**Checkoff for all PRs:**

 - [X] I have read the [Guidelines for Contributing](https://github.com/expfactory/expfactory/blob/master/.github/CONTRIBUTING.md) and this PR conforms to the stated requirements.
 - [ ] I have updated the [CHANGELOG](https://github.com/expfactory/expfactory/blob/master/CHANGELOG.md)

Attn: @expfactory-admin

Commit:
Merge pull request #1 from expfactory/add/https

Add/https

Commit:
Update 1-generate-https.md

Commit:
Update 1-generate-https.md

Commit:
Update 1-generate-https.md

Commit:
Update 1-generate-https.md

Commit:
Fix typo

vsoch commented on docs/pages/1-generate-https.md:
````diff
@@ -176,9 +176,9 @@ This will save all of the certificate files to /etc/letsencrypt/live/expfactory.
+cp /etc/letsencrypt/live/expfactory.dynu.net/fullchain.pem /etc/ssl/certs/chained.pem
+cp /etc/letsencrypt/live/expfactory.dynu.net/privkey.pem /etc/ssl/private/domain.key
+cp /etc/letsencrypt/ssl-dhparams.pem /etc/ssl/certs/dhparam.pem
 ```
````
most will definitely need sudo here! e.g.,

```
sudo cp <certificate> /etc/ssl/certs/<location>
```
quick question - is there a way to write the docs so that the letsencrypt path is variable, so the user could just copy it?

vsoch commented on docs/pages/1-generate-https.md:
````diff
@@ -132,32 +130,6 @@ sudo apt-get update
-```bash
-sudo reboot
-```
-
````
This section to add to Docker group and reboot is still needed.

tylerburleigh replied:
Is it? I believe the only reason we had to do this is because I didn't give you access to the root user. Instead I created a user for you to access my droplet. Normally the person creating the droplet will login as root, and may not have to follow these steps?

vsoch replied:
Yes, it definitely is. You can have sudo but once you are added to the docker daemon it needs to be restarted. Try it out - start a droplet having sudo (this was my case, for example), then try running docker. You will get a permission denied. if you add yourself to the group, you still get it. You would need to add then completely restart to get the right functionality.

Likely if you have seen otherwise, perhaps Digital Ocean takes the steps to add you to the docker group and then start the daemon in that state, so it only _seems_ like you "just need sudo."

tylerburleigh replied:
I get permission denied if I don't sudo, but I'm in the habit of sudo'ing everything. (Whether this is a good idea or not is a separate issue ;). I start docker with sudo and I issue any commands with sudo after that. I haven't had any issues doing this with DigitalOcean or with my AWS Bitnami LAMP stack.

vsoch replied:
We also don't want to assume that the user is necessarily root, or that it's even a droplet. If we have the step, worst case they are already able to use docker and they don't do it. If we don't have the step, the user might run into the error and not know what to do (and this would be poor documentation on our part for not anticipating this!)

vsoch replied:
It's a better idea to add the users to the docker group (and technically they don't need sudo, altough with docker they could escalate to it) vs. letting them use it by giving sudo. Generally, you don't want to be using sudo with Docker. With the docker group It should work without it.

Commit:
Tweaks to get it running in my server environment

Rewrite rules to run expfactory on /exp
Reverse proxy nginx to apache running outside docker

vsoch commented on script/nginx.https.conf:
```diff
@@ -22,8 +22,15 @@ server {
+    rewrite ^/save /exp/save last;
+    rewrite ^/static/(.*)$ /exp/static/$1 last;
+    rewrite ^/experiments/(.*)$ /exp/experiments/$1 last;
 
```
Hmm why are you doing this?

Commit:
Streamline copies w/ shell var

Commit:
Revert "Tweaks to get it running in my server environment"

This reverts commit 5f1871d1d459b770a4b1ac5d5431da79ff3e888d.

tylerburleigh committed:
Revert "Revert "Tweaks to get it running in my server environment""

This reverts commit bf4616fbbd5095d2f548806a37f55f8fb19e2dff.

tylerburleigh committed:
Revert "Revert "Revert "Tweaks to get it running in my server environment"""

This reverts commit 1896a06e0c7934c8649b25422fe2aa0f43085bb8.

vsoch commented on docs/pages/1-generate-https.md:
````diff
@@ -168,205 +140,66 @@ this is pronounced ENGINE-X - it took me only 8 years to know that :P). As a san
-   sudo cp /etc/ssl/certs/domain.csr{,.bak.$(date +%s)}
-fi
+certbot certonly --nginx -d expfactory.dynu.net -d www.expfactory.dynu.net
 ```
````
Is there a way to add an email, for notifications / other?

tylerburleigh replied:
Ah, I forgot to mention this, when certbot first generates a certificate, it asks for the user's email.
If you give it an email, it will send you a reminder at 20 days, 10 days, a 1 days before expiry.
https://letsencrypt.org/docs/expiration-emails/

tylerburleigh replied:
I've added this in.

vsoch replied:
is there a way to do this headless? We would want the user to be able to write a script to do it (e.g., not have any interactive prompts)

tylerburleigh replied:
We could pass a args to cert-bot to take care of the prompts

Instead of
`certbot certonly --nginx -d expfactory.dynu.net -d www.expfactory.dynu.net`

It could be 
`certbot certonly --nginx -d $DOMAIN1 -d $DOMAIN2 --email $EMAIL --agree-tos --redirect`

We would then have to also specify those shell vars above.

Is that all we need?

vsoch replied:
Yes! I like that. Let's take the approach that I was doing before, have the user define these variables once, and then use programmatically (that way they could be input arguments to a script). For example:

```bash
DOMAIN=expfactory.dynu.net
EMAIL=myawesome@email.com
```
```bash
certbot certonly --nginx -d "www.${DOMAIN}" -d "${DOMAIN}" --email "${EMAIL}" --agree-tos --redirect
```
We can assume (for now) that the user is just requesting certificates for the server he/she is sitting on (e.g., we wouldn't have domain1 and domain2.

vsoch replied:
yes that looks great!

Commit:
Add to Docker group

Commit:
Notes about renewal reminders, expiry

vsoch commented on docs/pages/1-generate-https.md:
````diff
@@ -168,205 +166,64 @@ this is pronounced ENGINE-X - it took me only 8 years to know that :P). As a san
+sudo add-apt-repository ppa:certbot/certbot
+sudo apt-get update
+sudo apt-get install python-certbot-nginx
 ```
````
Just a sanity check - do we not need to install python (e.g., does certbot (which I think based on the name needs python?) ) install it as a dependency? I needed to install python for Acme Tiny (the server doesn't come with it)

Commit:
Better use of shell vars

Commit:
Update 1-generate-https.md

vsoch commented on docs/pages/1-generate-https.md:
```diff
@@ -168,205 +166,64 @@ this is pronounced ENGINE-X - it took me only 8 years to know that :P). As a san
-### Step 5. Call Openssl
-We now are going to use openssl with our `csr_details.txt` to generate a new domain key! That looks like this:
+This will save all of the certificate files to /etc/letsencrypt/live/expfactory.dynu.net (the first domain listed in the cerbot command). It will ask for your email during the generation. This is used to send you renewal reminders at 20 days, 10 days, and 1 day before expiry (super helpful!)
 
```
oh hmm, so you are saying the above command generates a certificate for each domain? Is that a better approach than having one?

tylerburleigh replied:
My understanding is that it does, but it combines them into a single certificate file using the command above.

vsoch replied:
ah ok, so the user only sees one file. That works!

vsoch commented on docs/pages/1-generate-https.md:
````diff
@@ -168,205 +166,64 @@ this is pronounced ENGINE-X - it took me only 8 years to know that :P). As a san
-sudo mkdir -p /var/www/html/.well-known/acme-challenge/
-sudo chown $USER -R /var/www/html/
+```
+MYDOMAIN="expfactory.dynu.net"
````
Ah! Here is where the defined variable is handy, because they wouldn't need to define it again here.

vsoch commented on docs/pages/1-generate-https.md:
```diff
@@ -168,205 +166,67 @@ this is pronounced ENGINE-X - it took me only 8 years to know that :P). As a san
-sudo apt-get install -y python
+EMAIL="youremail@yourdomain.com"
+DOMAIN1="expfactory.dynu.net"
+DOMAIN2="www.expfactory.dynu.net"
```
Maybe just define `$DOMAIN` and have the command add the `www` ? (see my previous comment)

vsoch commented on docs/pages/1-generate-https.md:
````diff
@@ -168,205 +166,67 @@ this is pronounced ENGINE-X - it took me only 8 years to know that :P). As a san
 ```
 
-Then install Acme Tiny.
+The email you set here will be used to send you renewal reminders at 20 days, 10 days, and 1 day before expiry (super helpful!)
````
yes AMEN to those emails, I'd be lost without them!

Commit:
More streamlined shell vars

Commit:
Update 1-generate-https.md

Commit:
Update 1-generate-https.md

Commit:
Update 1-generate-https.md

Commit:
Step added for stopping nginx

vsoch commented:
Looks great! Merging.
